staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,616 @@
+		lastRealTime:     clk.Now(),
+		desiredNumQueues: desiredNumQueues,
+		concurrencyLimit: concurrencyLimit,
+		queueLengthLimit: queueLengthLimit,
+		requestWaitLimit: requestWaitLimit,
+		dealer:           dealer,
+	}
+	return fq, nil
+}
+
+// SetConfiguration is used to set the configuration for a queueSetImpl
+// update handling for when fields are updated is handled here as well -
+// eg: if desiredNumQueues is increased, SetConfiguration reconciles by
+// adding more queues.
+func (qs *queueSetImpl) SetConfiguration(concurrencyLimit, desiredNumQueues,
+	queueLengthLimit, handSize int, requestWaitLimit time.Duration) error {
+	// lock required as method can change Queues which has its indexes and length used
+	// concurrently
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+
+	dealer, err := shufflesharding.NewDealer(desiredNumQueues, handSize)
+	if err != nil {
+		return errors.Wrap(err, "shuffle sharding dealer creation failed")
+	}
+
+	// Adding queues is the only thing that requires immediate action
+	// Removing queues is handled by omitting indexes >desiredNumQueues from
+	// chooseQueueIdx
+
+	numQueues := len(qs.queues)
+	if desiredNumQueues > numQueues {
+		qs.queues = append(qs.queues,
+			initQueues(desiredNumQueues-numQueues, len(qs.queues))...)
+	}
+
+	qs.concurrencyLimit = concurrencyLimit
+	qs.desiredNumQueues = desiredNumQueues
+	qs.queueLengthLimit = queueLengthLimit
+	qs.requestWaitLimit = requestWaitLimit
+	qs.dealer = dealer
+
+	return nil
+}
+
+func removeRequest(slice []*fq.Request, s int) []*fq.Request {
+	return append(slice[:s], slice[s+1:]...)
+}
+
+// A quiescing QueueSet drains as normal but does not admit any
+// new requests. Passing a non-nil handler means the system should
+// become quiescent, a nil handler means the system should become
+//  non-quiescent. A call to Wait while the system is quiescent
+// will be rebuffed by returning `tryAnother=true`.  If all the
+// queues have no requests  waiting nor executing while the system
+// is quiescent then the handler will eventually be called with no
+// locks held (even if the system becomes non-quiescent between the
+// triggering state and the required call).
+//
+// The filter uses this for a priority level that has become
+// undesired, setting a handler that will cause the priority level
+// to eventually be removed from the filter if the filter still
+// wants that.  If the filter later changes its mind and wants to
+// preserve the priority level then the filter can use this to
+// cancel the handler registration.
+func (qs *queueSetImpl) Quiesce(eh fq.EmptyHandler) {
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+	if eh == nil {
+		qs.emptyHandler = eh
+		return
+	}
+	// Here we check whether there are any requests queued or executing and
+	// if not then fork an invocation of the EmptyHandler.
+	if qs.numRequestsEnqueued == 0 && qs.getRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		qs.counter.Add(1)
+		go func() {
```

Comment:
Shouldn't this `defer qs.counter.Add(-1)`?

Reply:
Oh yes, there is a bug here.  I think the right fix is for the forked func to end with `qs.counter.Add(-1)`.